fix(ngcc): implement lockfile

petebacondarwin opened the pull request:
Previously, it was possible for multiple instance of ngcc to be running
at the same time, but this is not supported and can cause confusing and
flakey errors at build time.

Now, only one instance of ngcc can run at a time. If a second instance
tries to execute it fails with an appropriate error message.

See https://github.com/angular/angular/issues/32431#issuecomment-571825781

petebacondarwin committed:
refactor(ivy): add `removeFile` to ngtsc `FileSystem`

petebacondarwin committed:
feat(ngcc): lock ngcc when processing

Previously, it was possible for multiple instance of ngcc to be running
at the same time, but this is not supported and can cause confusing and
flakey errors at build time.

Now, only one instance of ngcc can run at a time. If a second instance
tries to execute it fails with an appropriate error message.

See https://github.com/angular/angular/issues/32431#issuecomment-571825781

JoostK commented on packages/compiler-cli/ngcc/test/helpers/mock_lock_file.ts:
```diff
@@ -0,0 +1,27 @@
+export class MockLockFile extends LockFile {
+  log: string[] = [];
+  constructor(private options: {throwOnCreate?: boolean, throwOnRemove?: boolean} = {}) {
+    super(new MockFileSystemNative());
```
Could this be `getFileSystem` instead, as we run inside `runInEachFileSystem`?

petebacondarwin replied:
I have added a comment to the code. This `MockLockFile` is not used in tests that run `runInEachFileSystem`.

petebacondarwin committed:
refactor(ivy): add exclusive mode to `writeFile()`

This commit adds an `exclusive` parameter to the
`FileSystem.writeFile()` method. When this parameter is
true, the method will fail with an `EEXIST` error if the
file already exists on disk.

petebacondarwin commented on packages/compiler-cli/ngcc/src/execution/lock_file.ts:
```diff
@@ -0,0 +1,99 @@
+          `(If you are sure no ngcc process is running then you should delete the lockfile at ${this.lockFilePath}.)`);
+    }
+    this.fs.ensureDir(this.fs.dirname(this.lockFilePath));
+    this.fs.writeFile(this.lockFilePath, process.pid.toString());
```
I have changed the algorithm to use an exclusive write (and no exists check) which removes the change of a race condition, so I think we can get away with not adding the PID check here.

petebacondarwin commented on packages/compiler-cli/ngcc/src/execution/lock_file.ts:
```diff
@@ -0,0 +1,99 @@
+
+  protected addSignalHandlers() {
+    process.once('SIGINT', this.signalHandler);
+    process.once('SIGHUP', this.signalHandler);
```
No idea! I am hoping that our CI (or one of our Windows gurus) will test that for me :-)

petebacondarwin replied:
But to be serious, I looked into it and in theory this should be enough...

petebacondarwin commented on packages/compiler-cli/ngcc/src/execution/lock_file.ts:
```diff
@@ -0,0 +1,99 @@
+  protected remove() {
+    if (this.fs.exists(this.lockFilePath)) {
+      this.fs.removeFile(this.lockFilePath);
+    }
```
Sadly we can't remove the handlers here because it breaks the unit tests. This is because in the unit tests we don't actually kill the process so `remove()` still gets called :-/

petebacondarwin replied:
Hmm, but I just realised that the CLI will not exit the process after each ngcc run (i.e. they use `main.ts` directly. Which means that each time the CLI runs ngcc we will add more handlers...
So perhaps I do need to find a way to remove the handlers here - and still get it to work in the unit tests...

petebacondarwin replied:
And I can indeed run the `removeSignalHandlers()` here!

JoostK commented on packages/compiler-cli/ngcc/src/execution/lock_file.ts:
```diff
@@ -0,0 +1,116 @@
+      try {
+        pid = this.fs.readFile(this.lockFilePath);
+      } catch (e) {
+        pid = '{unkown}';
```
```suggestion
        pid = '{unknown}';
```

JoostK commented on packages/compiler-cli/ngcc/src/execution/lock_file.ts:
```diff
@@ -0,0 +1,116 @@
+      let pid: string;
+      try {
+        pid = this.fs.readFile(this.lockFilePath);
+      } catch (e) {
```
```suggestion
      } catch {
```

petebacondarwin replied:
Woh! I was sure that one had to provide the `(e)` even if you didn't use it. But I see that this is not the case. YLSED!

JoostK replied:
@petebacondarwin it's been supported since TS2.5! https://mariusschulz.com/blog/optional-catch-binding-in-typescript

petebacondarwin committed:
fix(ngcc): only lock ngcc after targeted entry-point check

The Angular CLI will continue to call ngcc on all possible packages, even if they
have already been processed by ngcc in a postinstall script.

In a parallel build environment, this was causing ngcc to complain that it was
being run in more than one process at the same time.

This commit moves the check for whether the targeted package has been
processed outside the locked code section, since there is no issue with
multiple ngcc processes from doing this check.

gkalpak commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -148,6 +149,20 @@ export function mainNgcc(
+          fileSystem, absoluteTargetEntryPointPath, supportedPropertiesToConsider,
+          compileAllFormats)) {
+    logger.debug('The target entry-point has already been processed');
+    return;
```
`mainNgcc` is expected to return a promise if `async` is true (returning `undefined` could break someone using it programatically and expecting a promise - not that there's anyone doing this, but better to be safe :grin:):

```suggestion
    return async ? Promise.resolve() : undefined;
```

gkalpak commented on packages/compiler-cli/ngcc/src/execution/lock_file.ts:
```diff
@@ -0,0 +1,117 @@
+ * When ngcc completes executing, it removes the file so that future ngcc executions can start.
+ */
+export class LockFile {
+  lockFilePath = this.fs.resolve(__dirname, '__ngcc_lock_file__');
```
Why not create it in a more "central" place, such as `compiler-cli/` or `compiler-cli/ngcc/`?
That would also be better DX imo (in case one needs to remove it manually after a crash or something).

petebacondarwin replied:
I did consider that but I was concerned about hard-coding the path to a more central place into this file. If folders/files get moved around then we would need to update the logic here, which IMO is brittle.

The path to the file is printed in the error message, so it is a simple cut-and-paste to delete it. And IMO there is not a huge difference in complexity for the developer between:

```
/Users/pete/dev/angular/tmp/ngcc-lockfile/integration/ngcc/node_modules/@angular/compiler-cli/ngcc/src/execution/__ngcc_lock_file__
```

and

```
/Users/pete/dev/angular/tmp/ngcc-lockfile/integration/ngcc/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__
```

JoostK replied:
@petebacondarwin Establishing the path of the ngcc root would be quite stable with `require.resolve('@angular/compiler-cli/ngcc')`?

petebacondarwin replied:
Done

gkalpak commented on packages/compiler-cli/ngcc/src/execution/lock_file.ts:
```diff
@@ -0,0 +1,117 @@
+      if (result instanceof Promise) {
+        isAsync = true;
+        // The cast is necessary because TS cannot deduce that T is now a promise here.
+        return result.finally(() => this.remove()) as unknown as T;
```
[According to MDN][1], `Promise.finally` is available in Node.js since v10. Should we play it safe and refrain from using it?

[1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally#Browser_compatibility

petebacondarwin replied:
The `package.json` for Angular indicates node.js >= 10.9; in CLI it is >=10.13. Moreover the CLI will actively error if the node.js version is below 10.13. See https://github.com/angular/angular-cli/blob/43e43bfacb7b59624b726ae1976e4d63d89bebf9/packages/angular/cli/lib/cli/index.ts#L26

So I think we are safe using this.

gkalpak replied:
The `package.json` at the repository root (which I assume is what you mean by "package.json for Angular") only applies during development (i.e. when one works in the repository). It does not affect the published packages (and their consumers). The [package.json][1] of the `compiler-cli` package currently requires Node.js >=8.

Also, while the latest CLI does require >=10.13, isn't it possible (although probably not recommended) that users could use the latest Angular with an older CLI version (or without the CLI at all)?

##
Maybe we just need to update `compiler-cli/package.json` to require Node.js v10 or newer. I think that would be best for v9.

[1]: https://github.com/angular/angular/blob/f2df1c79d/packages/compiler-cli/package.json#L31

petebacondarwin replied:
I agree that we should update the compiler-cli package.json. I'll add a commit to this PR.

petebacondarwin replied:
The last time it was changed was to align it with the CLI: https://github.com/angular/angular/commit/c602563589a7e14f0c57ee5b191979039aa370db

gkalpak commented on packages/compiler-cli/ngcc/src/execution/lock_file.ts:
```diff
@@ -0,0 +1,117 @@
+  protected signalHandler =
+      () => {
+        this.remove();
+        this.exit(1);
```
Given that `removeSignalHandlers` is now called in `this.remove()`, this could be just `process.exit(1)` (and the `exit()` method could be dropped).

petebacondarwin replied:
You are right, we can remove the call to `removeSignalHandlers()` from `exit()` but I want to keep that method because it is not possible to mock out `process.exit()` directly in unit tests: I found that the `process` object/namespace is not the same inside signal handlers - so if you replace `process.exit` in unit tests, the original one is still called inside the handler.

gkalpak commented on packages/compiler-cli/ngcc/src/execution/lock_file.ts:
```diff
@@ -0,0 +1,117 @@
+  lock<T>(fn: () => T): T {
+    let isAsync = false;
+    this.addSignalHandlers();
+    this.create();
```
So, if `this.create()` errors (and the process does not exit, because the caller handles the error), shouldn't signal handlers still be removed?

gkalpak commented on packages/compiler-cli/ngcc/src/execution/lock_file.ts:
```diff
@@ -0,0 +1,117 @@
+   * Write a lock file to disk, or error if there is already one there.
+   */
+  protected create() {
+    this.fs.ensureDir(this.fs.dirname(this.lockFilePath));
```
Is this necessary, given that it's `__dirname`?

petebacondarwin replied:
You are right that during normal operation this is not required. Good spot.

I threw this in because in many unit tests we are using a mock file-system that does not include the source files (and therefore doesn't contain this directory). If we don't do this here we have to change "a lot" of unit tests.

But I have just looked again, and I realise that I can add this to the `LockFileUnderTest` class rather than it appearing in the "real" `LockFile` class.

petebacondarwin replied:
Actually a whole load of the integration tests also need this folder to exist :-/

petebacondarwin replied:
I tweaked the integration tests to also ensure this folder exists...

gkalpak reviewed (changes requested):
Looking good! Some minor comments/questions/suggestions.

petebacondarwin committed:
feat(compiler-cli): require node 10 as runtime engine

Similar to c602563 this commit aligns the node.js version
requirements of the Angular compiler with Angular CLI.

gkalpak commented on packages/compiler-cli/ngcc/test/execution/lock_file_spec.ts:
```diff
@@ -0,0 +1,207 @@
+
+        lockFile.lock(() => {
+          lockFile.log.push('SIGINT');
+          process.emit('SIGINT', 'SIGINT');
```
OOC, why the second arg?

petebacondarwin replied:
I don't know if this is needed for the runtime but the typings for this overload (i.e. on that uses a signal string in the first parameter) looks like:

```ts
        emit(event: Signals, signal: Signals): boolean;
```

So at the very least, the second parameter is needed for TS compilation.

gkalpak reviewed (approved):
👌

matsko commented:
presubmit = https://test.corp.google.com/ui#id=OCL:290769792:BASE:290769840:1579630888550:8a26eed7

petebacondarwin commented:
@matsko - what is the reason for the presubmit failure?

AndrewKushnir commented:
@petebacondarwin it looks like there were some pre-existing failing tests. I marked "google3" check as passing for this PR. Thank you.

matsko commented:
@petebacondarwin please rebase.

petebacondarwin commented:
@matsko - done.
